allennlp/data/tokenizers/pretrained_transformer_tokenizer.py

```diff
@@ -177,139 +276,153 @@ def _tokenize(self, sentence_1: str, sentence_2: str = None):
-        Also calculates offsets such that wordpieces[offsets[i][0]:offsets[i][1] + 1]
+        Also calculates offsets such that tokens[offsets[i][0]:offsets[i][1] + 1]
         corresponds to the original i-th token.
 
         This function inserts special tokens.
         """
-        wordpieces, offsets, cumulative = self.intra_word_tokenize_in_id(
-            tokens, self.num_added_start_tokens
-        )
-        wp_tokens = self.ids_to_tokens(wordpieces)
-        assert cumulative + self.num_added_end_tokens == len(wp_tokens)
-        return wp_tokens, offsets
+        return self.intra_word_tokenize_sentence_pair(string_tokens, None)[:2]
 
     def intra_word_tokenize_sentence_pair(
-        self, tokens_a: List[str], tokens_b: List[str]
-    ) -> Tuple[List[Token], List[Tuple[int, int]], List[Tuple[int, int]]]:
+        self, string_tokens_a: List[str], string_tokens_b: Optional[List[str]]
+    ) -> Tuple[List[Token], List[Tuple[int, int]], Optional[List[Tuple[int, int]]]]:
         """
         Tokenizes each word into wordpieces separately and returns the wordpiece IDs.
         Also calculates offsets such that wordpieces[offsets[i][0]:offsets[i][1] + 1]
         corresponds to the original i-th token.
 
         This function inserts special tokens.
         """
-        wordpieces_a, offsets_a, cumulative = self.intra_word_tokenize_in_id(
-            tokens_a, self.num_added_start_tokens
-        )
-        wordpieces_b, offsets_b, cumulative = self.intra_word_tokenize_in_id(
-            tokens_b, cumulative + self.num_added_middle_tokens
-        )
-        wp_tokens = self.ids_to_tokens(wordpieces_a, wordpieces_b)
-        assert cumulative + self.num_added_end_tokens == len(wp_tokens)
-        return wp_tokens, offsets_a, offsets_b
 
-    def intra_word_tokenize_in_id(
-        self, tokens: List[str], starting_offset: int = 0
-    ) -> Tuple[List[int], List[Tuple[int, int]], int]:
-        """
-        Similar to `intra_word_tokenize()`, except:
-        (a) returns wordpiece IDs in the vocab instead of `Token`s;
-        (b) only takes a single sequence; and
-        (c) does not insert special tokens.
-        """
-        wordpieces: List[int] = []
-        offsets = []
-        cumulative = starting_offset
-        for token in tokens:
-            subword_wordpieces = self.tokenizer.encode(token, add_special_tokens=False)
-            if len(subword_wordpieces) == 0:
-                subword_wordpieces = [self.tokenizer.unk_token_id]
-
-            wordpieces.extend(subword_wordpieces)
-
-            start_offset = cumulative
-            cumulative += len(subword_wordpieces)
-            end_offset = cumulative - 1  # inclusive
-            offsets.append((start_offset, end_offset))
-        return wordpieces, offsets, cumulative
-
-    def ids_to_tokens(
-        self, token_ids_a: List[int], token_ids_b: Optional[List[int]] = None
+        def tokens_from_string_tokens(
```

Comment:
Do you really want to be defining functions inside the call to `intra_word_tokenize`?

Reply:
Why not? If I don't need the function anywhere else?

Reply:
It just seems like unfortunate overhead that happens in middle of a pre-processing loop.

Reply:
I refactored this, and I like the new version better. However, I always thought Python would JIT those functions once and never touch them again, so there is no overhead. So I tested it:

Before: `758 µs ± 943 ns per loop (mean ± std. dev. of 7 runs, 1000 loops each)`
After: `734 µs ± 2.21 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)`

So I guess it is 3% faster afterwards. Not a lot, but could matter in some cases. Interesting.

Reply:
Huh, thanks for measuring.  I did not assume anything would get JITed, so I thought there would be overhead, but I had no intuition for how large of a difference it would make.  Pretty small in the end, but if it's in the inner-loop of data processing, even small gains that increase GPU utilization are probably worth it.